services/src/datasets/external/natur40.rs

```diff
@@ -0,0 +1,290 @@
+use crate::{datasets::listing::DatasetListOptions, error::Result};
+use crate::{
+    datasets::{
+        listing::{DatasetListing, DatasetProvider},
+        storage::DatasetProviderDefinition,
+    },
+    error,
+    util::user_input::Validated,
+};
+use async_trait::async_trait;
+use futures::future::join_all;
+use gdal::DatasetOptions;
+use geoengine_datatypes::dataset::{DatasetId, DatasetProviderId, ExternalDatasetId};
+use geoengine_operators::engine::TypedResultDescriptor;
+use geoengine_operators::source::GdalMetaDataStatic;
+use geoengine_operators::util::gdal::{
+    gdal_open_dataset_ex, gdal_parameters_from_dataset, raster_descriptor_from_dataset,
+};
+use geoengine_operators::{
+    engine::{
+        MetaData, MetaDataProvider, RasterQueryRectangle, RasterResultDescriptor,
+        VectorQueryRectangle, VectorResultDescriptor,
+    },
+    mock::MockDatasetDataSourceLoadingInfo,
+    source::{GdalLoadingInfo, OgrSourceDataset},
+};
+use log::info;
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Natur40DataProviderDefinition {
+    id: DatasetProviderId,
+    name: String,
+    base_url: String,
+    user: String,
+    password: String,
+}
+
+#[typetag::serde]
+#[async_trait]
+impl DatasetProviderDefinition for Natur40DataProviderDefinition {
+    async fn initialize(self: Box<Self>) -> crate::error::Result<Box<dyn DatasetProvider>> {
+        Ok(Box::new(Natur40DataProvider {
+            id: self.id,
+            base_url: self.base_url,
+            user: self.user,
+            password: self.password,
+        }))
+    }
+
+    fn type_name(&self) -> String {
+        "Natur4.0".to_owned()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn id(&self) -> DatasetProviderId {
+        self.id
+    }
+}
+
+pub struct Natur40DataProvider {
+    id: DatasetProviderId,
+    base_url: String,
+    user: String,
+    password: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct RasterDb {
+    name: String,
+    title: String,
+    #[serde(deserialize_with = "string_or_string_array")]
+    tags: Vec<String>,
+}
+
+impl RasterDb {
+    fn url_from_name(base_url: &str, name: &str) -> String {
+        format!(
+            "WCS:{}/rasterdb/{name}/wcs?VERSION=1.0.0&COVERAGE={name}",
+            base_url,
+            name = name
+        )
+    }
+
+    fn url(&self, base_url: &str) -> String {
+        Self::url_from_name(base_url, &self.name)
+    }
+}
+
+#[derive(Deserialize, Debug)]
+struct RasterDbs {
+    rasterdbs: Vec<RasterDb>,
+    #[serde(deserialize_with = "string_or_string_array")]
+    tags: Vec<String>,
+    session: String,
+}
+
+#[async_trait]
+impl DatasetProvider for Natur40DataProvider {
+    async fn list(&self, _options: Validated<DatasetListOptions>) -> Result<Vec<DatasetListing>> {
+        // TODO: query the other dbs as well
```

Comment:
Which other dbs?

Reply:
I only query the rasterdb but there are others in the project that we have to integrate at some point

Reply:
Ah, and you would integrate them into this provider?

Reply:
I would imagine them all in this provider and the possibility to filter the external datasets by type etc. I think it would be worse having to look up project data in multiple places